CI: Overhaul for speed 🚀

sirreal opened the pull request:
This is a structural overhaul to the CircleCI setup. The goal was speed, leveraging CI parallelism and caching to make our CI runs as fast as possible 🚀 

Evidence is anecdotal, metrics from CI aren't great as far as I can tell, but `master` seems to run in 2-3 minutes. This branch's workflow runs 3 parallel jobs. `test-client`, the slowest job, tends to run in around a minute.

- Use shared references. This will helps to share common setup among jobs.
- Add eslint report to test summary. Because of eslint-eslines, we lose the inline report in exchange for the CI-based report.
- Inline CI test runners. The indirection through package.json or bins make it more difficult to work with the CI config. The indirection complicates things like using environment variables in the commands.
- Remove unneeded config and scripts now that CI test scripts are inlined.
- Cache the Jest cache.
- I explored caching the Babel cache, but that didn't seem to improve build time (e988d96fdb7b2ed8123b7d55c3c103ada2b2a4f1)
- I explored [`--changedSince=master`](https://jestjs.io/docs/en/cli#changedsince), but any difference was negligible. It's likely that the time taken to build out the dependency graph for `findRelatedTests` on every container negates any improvement.
- Add [`--ci`](https://jestjs.io/docs/en/cli#ci) Jest option
  > When this option is provided, Jest will assume it is running in a CI environment. This changes the behavior when a new snapshot is encountered. Instead of the regular behavior of storing a new snapshot automatically, it will fail the test and require Jest to be run with `--updateSnapshot`.
- Add `--runTestsByPath` Jest option for CircleCI split test. CircleCI splits a glob and passes full paths to Jest. Without this option, Jest will attempt to fuzzy match every full path it is passed.
  > `--runTestsByPath`
  > Used when provided patterns are exact file paths. This avoids converting them into a regular expression and matching it against every single file.
  > [boolean] [default: false]
- Add [`--silent `](https://jestjs.io/docs/en/cli#silent) Jest option
  > Prevent tests from printing messages through the console.
- Put test results directly into results folder for storage. Eliminates the need for a "gather results" step.
- Split sequential steps into a parallel workflow with different jobs:
  - `lint-and-translate`: Run eslint, generate translations. These jobs have been kept sequential because they appear to be fast enough that the overhead to startup 2 jobs is slower.
  - `test-client`: Run the client tests. This is the only job that seems worthwhile to run on parallel containers.
  - `test-server`: Run the server tests.
  - `test-integration`: This job is part of a cron workflow along with `test-client` and `test-server`. It is not run on demand (PR), but at scheduled intervals.
- Add `node_modules` cache. Previously we had eliminated the `node_modules` cache because we were using `npm ci` which deletes the folder (#25431). Then, we maintained this behavior, but began to cache the npm cache so that `npm ci` would execute faster because the modules were already available locally (#25487). This approach restores the `node_modules` cache and _completely skips_ `npm ci` if there was a cache hit. This should be safe because the `node_modules` cache requires a checksum match on `.nvmrc` and `npm-shrinkwrap.json`. It's important to ensure that _all variables_ which might generate different `node_modules` structure are part of the cache key, e.g. running on different architectures.
- Update the `git` cache with branch and revision. With a static key, the cache lives forever and is never updated.
- ~Add [`cache-require-paths`](https://www.npmjs.com/package/cache-require-paths) to the client and server Jest tests. In some anecdotal testing this seemed to provide improvement, but we might roll this back for now and do some more isolated testing after landing the bulk of this PR.~
- Quote shell variables

## Failing tests example

Here's an example of forced expected failure:

https://circleci.com/workflow-run/1ccbd0c0-19e2-4cc0-8de4-6ddfdf120d3a

![summary](https://user-images.githubusercontent.com/841763/45611698-98eb8a00-ba60-11e8-910b-f03f0ceb3c00.png)

Note that although lint fails, the sequential translate tasks still run.

This is the lint error summary that did not exist previously:

![lint-fail](https://user-images.githubusercontent.com/841763/45611668-76f20780-ba60-11e8-87c1-a3155fc76a3b.png)


## Testing
- Is CI working as expected?
- Do local test runs continue to work as expected?
  - `npm run test-client`
  - `npm run test-server`
  - `npm run test-integration`
- Is CI running faster? 😁 

## Follow-up
- Remove non CI-based cron to run integration tests
- `projects` based Jest running when monorepo begins (#27123)
- Pull e2e tests into this repo? This lays the groundwork…

From #27150

sirreal committed:
Restructure CI config with references

sirreal committed:
Inline CircleCI test runners

sirreal commented on .circleci/config.yml:
```diff
@@ -1,73 +1,121 @@
-          key: v1-20180524-git
-
+      ###############
+      # Basic Setup #
```
I'd love for this section to be shared, but YAML doesn't appear to allow merging arrays, so a sequence of steps cannot be merged like:

```yaml
steps:
  <<: *base-steps
  - another-step
```

ockham replied:
Hmm, really? The [docs](https://circleci.com/docs/2.0/writing-yaml/#merging-maps) seem to suggest that's possible, or am I misreading them?

sirreal replied:
🤔 hmmm…

So, we can't do my example, but maybe we can do like it suggests in that do. I understand we'd need to make another reference for the "main" steps of each job. I'll see if that works…

sirreal replied:
278b634 and f5992a5

This just doesn't seem to be possible.

sirreal commented on .circleci/config.yml:
```diff
@@ -81,18 +129,26 @@ jobs:
-              npm run test-server:ci $(circleci tests glob "server/**/test/*.js" "server/**/test/*.jsx" | circleci tests split --split-by=timings)
-
-      - run:
-          name: Gather Test Results
```
No longer necessary, just write test results to the appropriate locations.

sirreal commented on .circleci/config.yml:
```diff
@@ -1,73 +1,121 @@
+        "$CIRCLE_ARTIFACTS/translate"      \
+        "$CIRCLE_TEST_REPORTS/client"      \
+        "$CIRCLE_TEST_REPORTS/integration" \
+        "$CIRCLE_TEST_REPORTS/server"
```
CI will "categorize" test results. Each of these may have their own summary box if there's a report.

sirreal commented on .circleci/config.yml:
```diff
@@ -1,73 +1,121 @@
           name: Build calypso-strings.pot
           command: |
             if [ "$CIRCLE_NODE_INDEX" == "0" ]; then
-              npm run translate; mkdir -p $CIRCLE_ARTIFACTS/translate
```
We'll make this together with the rest of the directories above.

sirreal committed:
DROPME: force test failures to demonstrate reporting

sirreal committed:
DROPME: Demo workflow failures

sirreal committed:
Report eslint results

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,73 +1,110 @@
+                --reporters=jest-junit                     \
+                --config=test/integration/jest.config.json \
+                --maxWorkers=2                             \
+                $(circleci tests glob "bin/**/integration/*.js" "client/**/integration/*.js" "server/**/integration/*.js" | circleci tests split --split-by=timings)
```
for things like this how would having some scripts in a `Makefile` impact the reachability of the tests? does inlining them here make it harder to run the integration tests locally or is it totally independent of that?

sirreal replied:
> for things like this how would having some scripts in a `Makefile` impact the reachability of the tests?

I was hoping to hear some opinions about the inlining vs. npm script vs. bins vs. ?

I'm not very proficient in `make`, but all we need to do from this config file is to execute something from the shell and read its exit status (0/non-0).

I like the inlining because it gets rid of CI-related clutter around the repo and makes it easy to see what's happening right in the CI config.

> does inlining them here make it harder to run the integration tests locally or is it totally independent of that?

The tests continue to be available locally as `npm run test-{client,server,integration}`. They had `npm run test-{client,server,integration}:ci` variations that I don't think anyone would run locally, which I've removed in this PR.

sirreal commented on .circleci/config.yml:
```diff
@@ -1,73 +1,110 @@
+
+  restore-git-cache: &restore-git-cache
+    name: Restore git cache
+    key: v1-20180524-git
```
Just moved, but it's just a unique cache key.

I _think_ we should be using something more dynamic because CI caches are never overwritten, but I haven't included that in this PR.

dmsnell commented on .circleci/config.yml:
```diff
@@ -76,23 +113,55 @@ jobs:
+                | grep -E '^(client|server)'                          \
+                | grep -E '\.jsx?$'                                   \
+                | circleci tests split                                \
+              )                                                       \
```
fun fact, if we need to we can diff the output of two command and get the list of lines in one but not the other (if we need to)

```bash
man comm

comm -23 <(git diff --name-only | sort -u) <(git diff --name-only --cached | sort -u)
```

dmsnell commented on .circleci/config.yml:
```diff
@@ -76,23 +113,55 @@ jobs:
+            # Eslint would not produce a malformed empty xml using the --format and -o options.
+            # However, with eslint-eslines and the redirected > output, we may get an empty file.
+            # We only put the file in to the reports directory if the status code informs us that
+            # lint issues have been encountered.
```
lovely comment, sir

dmsnell commented on .circleci/config.yml:
```diff
@@ -76,23 +113,55 @@ jobs:
+            # We only put the file in to the reports directory if the status code informs us that
+            # lint issues have been encountered.
+            ESLINT_RESULT="$?"
+            if [[ "$ESLINT_RESULT" != "0" ]]; then
```
does this not work?

```bash
if [ $? -ne 0 ]; then
```

sirreal replied:
The `[ $VAR -ne 0 ]` bit should work (thanks), but we need to store the exit status in another variable so we can run other commands then exit with the status we're interested in (eslint's) so CI reads it. I may not have expressed that bit well in the comment.

sirreal committed:
Try changedSince

sirreal committed:
Use js for jest config

sirreal committed:
Persist Jest caches between builds

sirreal committed:
Better git caching

sirreal committed:
Position Jest tests in adjacently

sirreal committed:
Remove "dev" from cache keys

sirreal committed:
Revert "DROPME: force test failures to demonstrate reporting"

This reverts commit f865023d8f78b5a4247e44f24d4980efc516f848.

sirreal committed:
Revert "Try changedSince"

This reverts commit 8029fd9f923c25a412d7e9d4643d4e27c0f5369d.

sirreal commented:
2:15 No git/jest cache build with `changedSince='master'`: https://circleci.com/gh/Automattic/wp-calypso/104352
1:43 Cached build with `changedSince='master'`: https://circleci.com/gh/Automattic/wp-calypso/104354

sirreal commented:
2:10 - For reference, this is the faster recent `master` build: https://circleci.com/gh/Automattic/wp-calypso/104328

1:33 - Without `changedSince`: https://circleci.com/gh/Automattic/wp-calypso/104355
1:37 - Without `changedSince`: https://circleci.com/gh/Automattic/wp-calypso/104356

sirreal committed:
Use a parallel CI workflow

sirreal committed:
Isolate test-client for optimization

sirreal committed:
Optimize test-client

sirreal committed:
Cache node_modules

sirreal committed:
Reenable tests and further split

sirreal committed:
Remove npx invocation for speed

sirreal committed:
Name workflow "Calypso"

sirreal committed:
Split by timings

sirreal committed:
Bump jest cache version (timings)

sirreal committed:
Remove Jest cache from integration

sirreal committed:
Shift parallelism to client

sirreal committed:
Run integration nightly

sirreal committed:
Use 4 client workers

sirreal commented:
This seems to be in a good place. The entire workflow (translate, lint, client, server) runs in about 1 minute. The integration tests are scheduled to run via cron.

There's some more tweaking that could be done – parallelization etc.
- The client could use a bit more parallelization. 6 seems to be pretty good. It never gets much below 1 minute with caches primed, but at 4 it suffers with no cache.
- The jobs that aren't parallel can have their splitting removed.
- Lint only runs on a few files and doesn't need its own job. It could run in the same job as translate to save some boot overhead.
- The server tests are also quite fast and could potentially also join lint and translate.
- The Jest workers seem pretty good at 2. That makes some sense since we probably have 2 cores (https://circleci.com/docs/2.0/configuration-reference/#resource_class)

sirreal committed:
Update nightly name and suites

sirreal committed:
Cache requires when running Jest

sirreal committed:
Remove CI splitting for non-parallel jobs

sirreal committed:
Use only node_modules cache

sirreal committed:
Add babel cache cache

sirreal committed:
Revert "Add babel cache cache"

This reverts commit e988d96fdb7b2ed8123b7d55c3c103ada2b2a4f1.

sirreal committed:
Merge lint and translate jobs

These are faster tasks that are faster to run sequentially than in
parallel do to job setup overhead.

sirreal committed:
Test results by directory not filename

sirreal committed:
Split by filename

Timings may change, filenames should be more stable.

Ensure Jest caches by container match files

sirreal committed:
Try timings again with cache-bust

sirreal committed:
Add 30s timeout to Jest

I've just observed a hung test suite that waited
10 minutes. Let's ensure that doesn't happen again.

sirreal committed:
Increase timeout to 2m

An in-progress run of the server test timed out after 30s. Increase the
timeout to 2m to avoid timing out in-progress runs.

sirreal committed:
Add changedSince again

sirreal committed:
Drop changedSince

sirreal committed:
Revert "DROPME: Demo workflow failures"

This reverts commit f27a84dd341202adca02c4580e7cd7163e0126d7.

sirreal committed:
Move cacheDirectory option out of config

sirreal committed:
Bump jest cache

sirreal commented:
I'd love to get this reviewed now!

ockham commented on .circleci/config.yml:
```diff
@@ -1,109 +1,303 @@
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}-{{ .Revision }}
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-master
+      - v6-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}
```
I see we're storing Jest cache using the `-{{ .Branch }}-{{ .Revision }}` key below, but where are we using the others?

sirreal replied:
These are the restores, so CircleCI will try to match from first to last until it gets a hit or fails and doesn't restore a cache.

- Ideally, we have a cache from this revision.
- Otherwise, from this branch
- Otherwise, from master
- …

---

The `v*` prefixes are just ways of cache busting. CircleCI caches do not necessarily expire. As I've iterated on this I've used several, which is how we got a relatively high number. Given that caches are never overwritten, if something needs to change in the way things are cached, it's a good idea to cache bust by bumping this.

blowery replied:
> Otherwise, from master

This isn't really master though. It's whatever master was the first time we had a run for master. I'm not sure it's really a good fallback...

sirreal replied:
> This isn't really master though.

It _could_ be master, but mostly it won't 🙂.  master probably has a decent cache we can leverage rather than starting from scratch. We're just warming up the Jest cache, Jest will be able to leverage some of the cache from master, and what it doesn't use I wouldn't expect to be harmful.

> It's whatever master was the first time we had a run for master.

It should be from the HEAD, unless we somehow generate a cache key from an old commit on the master branch. Master caches will have keys like

```
v6-jest-test-client-0/5-master-abcd123
```

A new cache is created for each commit on each branch, and we default to the [most recent one](https://circleci.com/docs/2.0/caching/#restoring-cache):

> The cache will be restored from the first matching key. If there are multiple matches, the most recently generated cache will be used.

If PRs aren't "far" from master, this is probably a good way to warm the cache.

Does that seem like sound reasoning?

blowery replied:
ah, interesting, I missed the most-recent-cache bit on restores. Cool!

ockham commented on .circleci/config.yml:
```diff
@@ -1,109 +1,303 @@
+            ESLINT_RESULT="$?"
+            if [ "$ESLINT_RESULT" -ne 0 ]; then
+              mv ~/eslint-results.xml "$CIRCLE_TEST_REPORTS/eslint/results.xml"
+              exit "$ESLINT_RESULT"
```
Do we need to delete `~/eslint-results.xml` after this?

sirreal replied:
If there is a report, we move the file to the test_results so it will be uploaded and we need to keep it.

If there isn't a report, it shouldn't matter. These containers are ephemeral and `eslint-results.xml` will wind in the ether where all deleted files go.

ockham commented on .circleci/config.yml:
```diff
@@ -1,109 +1,303 @@
+      - store_artifacts:
+          path: /tmp/test_results
+      - store_artifacts:
+          path: /tmp/artifacts
```
Why are there two `store_artifacts` (with different `path`s) specified?

sirreal replied:
This uploads different things associated with the CI job. We can likely rework this so that both directories can be uploaded in a single step. The artifacts contain output from translation. Uploading the test results just makes the whole test report available for someone to access at a later point. This existed before and hasn't really changed with this PR.

sirreal committed:
Add docs describing the cache mechanisms

sirreal committed:
Avoid split-by timings for Jest splitting

sirreal commented on test/client/setup-test-framework.js:
```diff
@@ -1,10 +1,8 @@
 /**
  * External dependencies
  */
+import 'cache-require-paths';
```
> Add [`cache-require-paths`](https://www.npmjs.com/package/cache-require-paths) to the client and server Jest tests. In some anecdotal testing this seemed to provide improvement, but we might roll this back for now and do some more isolated testing after landing the bulk of this PR.

Thoughts?

sirreal replied:
Rolled it back 99ce096263227cc45922fbfdbc399ff714951f5b

sirreal commented on test/server/setup-test-framework.js:
```diff
@@ -1,6 +1,8 @@
 /**
  * External dependencies
  */
+import 'cache-require-paths';
```
> Add [`cache-require-paths`](https://www.npmjs.com/package/cache-require-paths) to the client and server Jest tests. In some anecdotal testing this seemed to provide improvement, but we might roll this back for now and do some more isolated testing after landing the bulk of this PR.

Thoughts?

sirreal committed:
Reuse merged step arrays to build jobs

sirreal committed:
Revert "Reuse merged step arrays to build jobs"

This reverts commit 278b63455157deed2e074a03ce6503a20357c9f3.

sirreal committed:
Rollback cache-require-paths addition

alisterscott commented:
This is an awesome change that looks great - the only thing I saw in the results was the message about the translate hook in each build results:

`webhook returned http status 400, connecting to https://translate.wordpress.com/api/localci/-relay-new-strings-to-gh
`

Is this expected?

sirreal commented:
> Is this expected?

No. I was seeing this issue intermittently, but it seems to be occurring regularly now. I'll try to have a look, but I'd love a hand from @yoavf or @hewsut if they have a moment to look (they seem to have worked on the webhook API).

sirreal commented:
Translations conversation: p1537262192000100-slack-i18n

sirreal commented:
I've prepped a candidate to improve the notify webhook handling:

D18455-code

blowery commented:
this is amazing!

sirreal committed:
Replace CircleCI v1 notify with explicit curl

sirreal committed:
DROPME: Testing translation notify/fetch

sirreal commented:
It appears that the notify step is working well 🎉 

p1537515574000100-slack-i18n-debug

If folks are happy, I'd like to get this formally approved and land it 🚀

sirreal commented:
I'm proposing a fix for the last issues I found with GlotPress and CircleCI. There were some hidden problems 🙂

sirreal commented:
This is blocked by D18656-code. Otherwise, it's ready.

Before merging, we need to revert the demo translation change (aff4202).

Optionally, we can remove this branch from the cron filter.

blowery reviewed (approved):
LGTM once you have the i18n bits sorted out!

Lovely work @sirreal! Thanks for taking this on.

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,109 +1,359 @@
+    name: Restore Jest cache
+    keys:
+      - v7-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}-{{ .Revision }}
+      - v7-jest-{{ .Environment.CIRCLE_JOB }}-{{ .Environment.CIRCLE_NODE_INDEX }}/{{ .Environment.CIRCLE_NODE_TOTAL }}-{{ .Branch }}
```
since branches can be rebased and can change will this cause cache pollution? in other words, should be caching on the commit hash instead of the branch head?

sirreal replied:
I wouldn't expect it to cause problems. This is warming the Jest cache. It will try to match a cache from the exact revision, then the most recent from the branch, then the most recent from master…

The most recent cache from the same branch is probably a good guess. In the case of the first commit on an old branch after rebasing on master, it would be better to match master, but I'd expect that to be a less common case.

I'm not completely certain, but I don't think there's any danger here. We just have more or less warm caches. I expect the same situation arises with local test runs. After rebasing, Jest will have a cache based on the older branch point. That doesn't mean the tests become inaccurate, just that Jest needs to do more work than with a very recent cache.

Related: https://github.com/Automattic/wp-calypso/pull/27180#discussion_r218834521

dmsnell replied:
I'm thinking about this for all these `.branch` caches as well. Just to confirm:

 - I create a branch and push code to it
 - CircleCI writes to these caches for my branch
 - I push a new commit to my branch (not even rebasing)
 - CircleCI reuses the old cache

Or does CircleCI even use the branch tip when it uses `.Branch` - I guess it's possible that the syntax is shorthand for "current branch tip" vs. "current branch name"

sirreal replied:
> * I create a branch and push code to it
> * CircleCI writes to these caches for my branch
> * I push a new commit to my branch (not even rebasing)
> * CircleCI reuses the old cache

Yes and no 🙃 These cache restores will cascade, matching the first that hits, top to bottom. In this case, on the first build:

- Attempt restore `…-{{ .Branch }}-{{ .Revision }}`. Cache miss
- Attempt restore `…-{{ .Branch }}`. All revisions from this branch hit. The most recent cache from this branch will be restored.
- Jest runs tests…
- `…-{{ .Branch }}-{{ .Revision }}` New Jest cache cache is stored.

On subsequent runs of this revision, we'll match the new Jest cache cache that matches exactly. Also, the next commit to this branch will match this newly created cache.

---

I don't understand the fundamental concern. We are not caching test results. Jest still runs, we only restore Jest's cache so that it's warm rather than cold. I would not expect test results to be inaccurate due to an older cache being provided for Jest. As I understand it, that's the purpose and expected behavior of a cache like the one Jest uses– to store some previous work that's already been done to avoid repeating it.

dmsnell replied:
According to the docs `.Revision` is `The VCS revision currently being built.` which I guess means commit hash…?

My fundamental concern is restoring files from a cache that have updated and then we "avoid repeating" those tests because we think they didn't change.

I'm only trying to make sure we don't skip tests due to restoring a cache from old code.

sirreal replied:
> According to [the docs](https://circleci.com/docs/2.0/caching/) `.Revision` is `The VCS revision currently being built.` which I guess means commit hash…?

Yep 👍 

> My fundamental concern is restoring files from a cache that have updated and then we "avoid repeating" those tests because we think they didn't change.
> 
> I'm only trying to make sure we don't skip tests due to restoring a cache from old code.

I'm confident that is not happening. Jest does have options to skip tests based on changes, but we're not using any of them. I've pushed a commit to demonstrate how Jest fails correctly with this caching scheme: https://circleci.com/workflow-run/8de0cfa1-4ae0-45b6-bd77-30c607012907

dmsnell replied:
your confidence gives me confidence 😄

sirreal committed:
DROPME: Force test failures

sirreal committed:
Revert "DROPME: Force test failures"

This reverts commit bcf6ea074af217fea78c38466bc23179aa441550.

sirreal committed:
Revert "DROPME: Testing translation notify/fetch"

This reverts commit aff4202b2b420552e0e5697615cf02299079143a.

sirreal committed:
Run nightly tests on master branch only
